flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/hybrid/HybridSource.java

```diff
@@ -0,0 +1,163 @@
+        for (Tuple2<Source<T, ? extends SourceSplit, ?>, ?> source : sourceChain.sources) {
+            readers.add(source.f0.createReader(readerContext));
+        }
+        return new HybridSourceReader(readerContext, readers);
+    }
+
+    @Override
+    public SplitEnumerator<HybridSourceSplit, HybridSourceEnumeratorState> createEnumerator(
+            SplitEnumeratorContext<HybridSourceSplit> enumContext) {
+        return new HybridSourceSplitEnumerator(enumContext, sourceChain);
+    }
+
+    @Override
+    public SplitEnumerator<HybridSourceSplit, HybridSourceEnumeratorState> restoreEnumerator(
+            SplitEnumeratorContext<HybridSourceSplit> enumContext,
+            HybridSourceEnumeratorState checkpoint)
+            throws Exception {
+        return new HybridSourceSplitEnumerator(
+                enumContext, sourceChain, checkpoint.getCurrentSourceIndex());
+    }
+
+    @Override
+    public SimpleVersionedSerializer<HybridSourceSplit> getSplitSerializer() {
+        List<SimpleVersionedSerializer<SourceSplit>> serializers = new ArrayList<>();
+        sourceChain.sources.forEach(
+                t -> serializers.add(castSerializer(t.f0.getSplitSerializer())));
+        return new HybridSourceSplitSerializer(serializers);
+    }
+
+    @Override
+    public SimpleVersionedSerializer<HybridSourceEnumeratorState>
+            getEnumeratorCheckpointSerializer() {
+        List<SimpleVersionedSerializer<Object>> serializers = new ArrayList<>();
+        sourceChain.sources.forEach(
+                t -> serializers.add(castSerializer(t.f0.getEnumeratorCheckpointSerializer())));
+        return new HybridSourceEnumeratorStateSerializer(serializers);
+    }
+
+    private static <T> SimpleVersionedSerializer<T> castSerializer(
+            SimpleVersionedSerializer<? extends T> s) {
+        @SuppressWarnings("rawtypes")
+        SimpleVersionedSerializer s1 = s;
+        return s1;
+    }
+
+    /**
+     * Converts checkpoint between sources to transfer end position to next source's start position.
+     * Only required for dynamic position transfer at time of switching, otherwise source can be
+     * preconfigured with a start position during job submission.
+     */
+    public interface CheckpointConverter<InCheckpointT, OutCheckpointT>
+            extends Function<InCheckpointT, OutCheckpointT>, Serializable {}
+
+    /** Chain of sources with option to convert start position at switch-time. */
+    public static class SourceChain<T, SplitT extends SourceSplit, EnumChkT>
+            implements Serializable {
+        final List<Tuple2<Source<T, ? extends SourceSplit, ?>, CheckpointConverter<?, ?>>> sources;
+
+        public SourceChain(Source<T, SplitT, EnumChkT> initialSource) {
+            this(concat(Collections.emptyList(), Tuple2.of(initialSource, null)));
+        }
+
+        public static <T, SplitT extends SourceSplit, EnumChkT> SourceChain<T, SplitT, EnumChkT> of(
```

Comment:
Does the `of` method need the `CheckpointConverter`?

Reply:
This method is a shortcut for fixed start position. Will add javadoc.